sentinel-adapter/sentinel-apache-dubbo-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/dubbo/SentinelDubboConsumerFilter.java

```
@@ -52,32 +52,39 @@ public SentinelDubboConsumerFilter() {
+                methodEntry = SphU.entry(resourceName, EntryType.OUT);
+            } else {
+                // should generate the AsyncEntry when the invoke model in future or async
+                interfaceEntry = SphU.asyncEntry(invoker.getUrl().getEncodedServiceKey(), EntryType.OUT);
+                methodEntry = SphU.asyncEntry(resourceName, EntryType.OUT);
+            }
+            rpcContext.set(DubboUtils.DUBBO_INTERFACE_ENTRY_KEY, interfaceEntry);
+            rpcContext.set(DubboUtils.DUBBO_METHOD_ENTRY_KEY, methodEntry);
             Result result = invoker.invoke(invocation);
-            if (result.hasException()) {
-                Throwable e = result.getException();
-                // Record common exception.
-                Tracer.traceEntry(e, interfaceEntry);
-                Tracer.traceEntry(e, methodEntry);
+            if (result instanceof AsyncRpcResult) {
+                // catch timeout or nonbiz-exception when in async model
+                AsyncRpcResult asyncRpcResult = (AsyncRpcResult) result;
+                asyncRpcResult.getValueFuture().whenComplete((rs,ex) -> {
+                    if (ex != null){
+                        trace((Throwable) ex, invocation);
+                    }
+                });
             }
             return result;
         } catch (BlockException e) {
             return DubboFallbackRegistry.getConsumerFallback().handle(invoker, invocation, e);
         } catch (RpcException e) {
-            Tracer.traceEntry(e, interfaceEntry);
-            Tracer.traceEntry(e, methodEntry);
+            // catch timeout or nonbiz-exception when in sync model
+            trace(e, invocation);
             throw e;
-        } finally {
-            if (methodEntry != null) {
-                methodEntry.exit();
-            }
-            if (interfaceEntry != null) {
-                interfaceEntry.exit();
-            }
         }
```

`interfaceEntry` and `interfaceEntry`  exit() method should ensure  executed  regardless of whether there is an exception?

---

The logic might be:

- When the invocation is successful, then the entries will be exited in `onResponse` callback.
- When the invocation fails, then record the exception and complete the entries. For asynchronous call, it's handled in `whenComplete` callback.

But for the synchronous case, we may need to check whether the result has exceptions or not, or some business exceptions might be ignored.

Besides, could you please provide a more detailed design in the PR description? @CodingSinger

---

根据我的理解，在Dubbo2.7.1中，请求如果正常返回，则会回调`Filter#onResponse`,如果请求超时或者是发生Rpc框架异常，则分两种情况：
1. 当同步调用时，需要自己try catch来捕获。
2. 当异步调用时，异常会在filter丢失，如果需要在filter记录异常，需要自己用`whenComplete`的方式来回调。

综上，我们需要在三个地方埋点。onResponse的trace针对正常返回，`try...catch`的trace针对同步调用异常，`whenComplete`针对异步调用异常